rfcs/proposed/parallel_block_for_task_arena/README.md

```diff
@@ -0,0 +1,189 @@
+where parallel computation ends, they can also indicate where it starts.
+
+<img src="parallel_block_introduction.png" width=800>
+
+With this approach, the user not only releases threads when necessary but also specifies a
+programmable block where worker threads should expected new work coming regularly
+to the executing arena.
+
+Let’s add new state to the existing state machine. To represent "Parallel Block" state.
+
+> **_NOTE:_** The "Fast leave" state is colored Grey just for simplicity of the chart.
+              Let's assume that arena was created with the "Delayed leave". 
+              The logic demonstrated below is applicable to the "Fast leave" as well.
+
+<img src="parallel_block_state_initial.png" width=800>
+
+This state diagram leads to several questions. There are some of them:
+* What if there are multiple Parallel Blocks?
+* If “End of Parallel Block” leads back to “Delayed leave” how soon threads
+  will be released from arena?
+  * What if we indicated that threads should leave arena after the "Parallel Block"?
+  * What if we just indicated the end of the "Parallel Block"?
+
+The extended state machine aims to answer these questions.
+* The first call to the “Start of PB” will transition into the “Parallel Block” state.
+* The last call to the “End of PB” will transition back to the “Delayed leave” state
+  or into the "One-time Fast leave" if it is indicated that threads should leave sooner.
+* Concurrent or nested calls to the “Start of PB” or the “End of PB”
+  increment/decrement reference counter.
+
+<img src="parallel_block_state_final.png" width=800>
+
+Let's consider the semantics that an API for explicit parallel blocks can provide:
+* Start of a parallel block:
+  * Indicates the point from which the scheduler can use a hint and keep threads in the arena
+    for longer.
+  * Serves as a warm-up hint to the scheduler:
+    * Allows worker threads to be available by the time real computation starts.
+* "Parallel block" itself:
+  * Scheduler can implement different policies to retain threads in the arena.
+  * The semantics for retaining threads is a hint to the scheduler;
+    thus, no real guarantee is provided. The scheduler can ignore the hint and
+    move threads to another arena or to sleep if conditions are met.
+* End of a parallel block:
+  * Indicates the point from which the scheduler may drop a hint and
+    no longer retain threads in the arena.
+  * Indicates that arena should enter the “One-time Fast leave” thus workers can leave sooner.
+    * If work was submitted immediately after the end of the parallel block,
+      the default arena "workers leave" state will be restored.
+    * If the default "workers leave" state was the "Fast leave" the result is NOP.
+
+
+### Proposed API
+
+```cpp
+class task_arena {
+    enum class workers_leave : /* unspecified type */ {
+        fast = /* unspecifed */,
+        delayed = /* unspecifed */
+    };
```

Comment:
We need to find better names to the enum class and its values.
I am specifically concerned about the use of "delayed" in case the actual behavior might be platform specific, not always delayed. But also `workers_leave` is not a very good name.

Reply:
I've been thinking... Combined with your previous comment about `automatic`, perhaps we could have 3 modes instead:
* automatic - platform specific default setting
* fast (or any other more appropriate name)
* delayed (or any other more appropriate name)

If we assume that we have these 3 modes now, `fast` and `delayed` modes would enforce behavior regardless of the platform. That would allow to use `start/end_parallel_block` on hybrid CPUs but if we set `automatic` instead, it would be translated to `fast` option for hybrid systems (to be aligned with current implementation of block time here https://github.com/uxlfoundation/oneTBB/blob/master/src/tbb/waiters.h#L61).
What do you think?

Reply:
Yes, that's an option we can consider. We'd need to think though what the "enforced" delayed mode will guarantee.

For the "automatic" mode, we say that after work completion threads might be retained in the arena for unspecified time chosen by internal heuristics. How would the definition of "enforced" delayed mode differ, what additional guarantees it would provide to make sense for users to choose it?

Reply:
Seems like if we have a dedicated `automatic` policy, that means that `delayed` policy should guarantee at least some level of threads' retention relatively to `fast` policy.

Reply:
I think the whole structure is just a hint to the scheduler no real guarantees provided therefore from the description we get:
1) threads will leave without delay with "Fast" 
2) threads might have a delay before leaving with "Delayed" 
3) automatic will decide what state to choose 

From the implementation stand point it has a lot of sense since we will have clear invariants for arena e.g., default arena on hybrid platform will have "Fast" leave state.
So it defiantly improves implementation logic while bringing some potential value to the users ("Delayed" will behave as thread retantion if user explicitly specified).

Reply:
@akukanov Regarding enumeration class name. Do you find `leave_policy` a better name than `workers_leave`? Seems more natural to me when using it during arena construction:
```cpp
tbb::task_arena ta{..., tbb::task_arena::leave_policy::fast};
```

Reply:
@pavelkumbrasev please describe the semantics of the `delayed` hint in a way that is meaningful for users.

For example, the description I used above 
> after work completion threads might be retained in the arena for unspecified time chosen by internal heuristics.

as well as what you mentioned
> threads might have a delay before leaving with "Delayed"

are good for `automatic` but rather bad for `delayed` because all aspects of the decision are left to the implementation. Even if changed to "will be retained for unspecified time", it would still be rather weak because the time can be any, including arbitrarily close to 0 - that is, it's not really different from `automatic` and so there is no reason to prefer it.

Reply:
Sure. Perhaps we are not on the same page. I would like it to be:
1. delayed - after work completion threads might be retained in the arena for unspecified time chosen by internal heuristics.
2. automatic - implementation will choose between "fast" and "delayed"

Automatic is basically another heuristic on choosing "fast" or "leave" based on underlying HW. Perhaps, automatic not the best name.

Reply:
With this definition, I see no difference between "automatic" and "delayed", because in both cases the decision of whether to delay or not to delay, and for how long, is left to the implementation. If that is the intended behavior, let's not complicate the API with a redundant enum value.

Reply:
Okay, let's have `fast` and `automatic` policies then since we're not sure right now whether we can provide some meaningful guarantees for `delayed` policy to user.